pkg/tagger/collectors/containerd_main.go

```diff
@@ -0,0 +1,240 @@
+	_ Streamer  = &ContainderdCollector{}
+	_ Fetcher   = &ContainderdCollector{}
+)
+
+// Detect tries to connect to the containerd socket and initializes the ContainderdCollector attributes.
+// Fast return if containerd wasn't detected by env discovery.
+func (c *ContainderdCollector) Detect(_ context.Context, out chan<- []*TagInfo) (CollectionMode, error) {
+	if !config.IsFeaturePresent(config.Containerd) {
+		return NoCollection, nil
+	}
+
+	client, err := cutil.GetContainerdUtil()
+	if err != nil {
+		return NoCollection, err
+	}
+
+	c.client = client
+	c.infoOut = out
+	c.stop = make(chan struct{})
+	c.envAsTags, c.globEnv = utils.InitMetadataAsTags(config.Datadog.GetStringMapString("container_env_as_tags"))
+
+	return StreamCollection, nil
+}
+
+// Stream watches and processes containers lifecycle events.
+func (c *ContainderdCollector) Stream() error {
+	healthCtx, healthCancel := context.WithCancel(context.Background())
+	health := health.RegisterLiveness("tagger-containerd")
+
+	streamCtx, streamCancel := context.WithCancel(context.Background())
+	stream, streamErrors := c.client.GetEvents().Subscribe(streamCtx, containerLifecycleFilters...)
+
+	for {
+		select {
+		case <-c.stop:
+			healthCancel()
+			streamCancel()
+			return health.Deregister()
+		case healthDeadline := <-health.C:
+			healthCancel()
+			healthCtx, healthCancel = context.WithDeadline(context.Background(), healthDeadline)
+		case event := <-stream:
+			c.processEvent(healthCtx, event)
+		case err := <-streamErrors:
+			if err == nil {
+				continue
+			}
+
+			log.Errorf("Containerd events stream error: %w", err)
+			healthCancel()
+			streamCancel()
+
+			return err
+		}
+	}
+}
+
+// Fetch tries to get tags for a specific container by its ID.
+func (c *ContainderdCollector) Fetch(ctx context.Context, entity string) ([]string, []string, []string, error) {
+	entityType, id := containers.SplitEntityName(entity)
+	if entityType != containers.ContainerEntityName || len(id) == 0 {
+		return nil, nil, nil, nil
+	}
+
+	container, isPauseContainer, err := c.containerByID(ctx, id)
+	if err != nil {
+		log.Debugf("Could not fetch container %s - %w", id, err)
+		return nil, nil, nil, err
+	}
+
+	if isPauseContainer {
+		log.Debugf("Ignoring pause container %s", id)
+		return nil, nil, nil, nil
+	}
+
+	low, orchestrator, high, _, err := c.tagsForContainer(ctx, container)
+	return low, orchestrator, high, err
+}
+
+// Stop shut down the container events watching loop.
+func (c *ContainderdCollector) Stop() error {
+	close(c.stop)
+	return nil
+}
+
+// processEvent handles container creation/update/deletion events and sends a TagInfo accordingly.
+func (c *ContainderdCollector) processEvent(ctx context.Context, event *containerdevents.Envelope) {
+	var handleCreateUpdate = func(id string) {
+		container, isPauseContainer, err := c.containerByID(ctx, id)
+		if err != nil {
+			log.Debugf("Could not fetch container %q: %w", id, err)
+			return
+		}
+
+		if isPauseContainer {
```

Comment:
I wonder if we shouldn't emit a `TagInfo` with no tags for such cases, as a `tagger.Tag` call about a pause container will always result in a fetch (which is relatively expensive) that never returns anything, although we know that a pause container will never have any tags.

Reply:
Why would we call `tagger.Tag` with an entity corresponding to to a pause container? Are you thinking of a particular tagger client that could do it?

Reply:
No, I don't have a particular one in mind. Ideally they shouldn't, I agree. Just monitoring for spikes in entities not found and fixing it at call sites is a reasonable way to go, too, so all in all this LGTM as it is 🚀